cmd/mirror/mirror.go

```diff
@@ -0,0 +1,484 @@
+	if err != nil {
+		return nil, err
+	}
+	ret := make(map[string]listedModule)
+	var jsonBuilder strings.Builder
+	for _, line := range strings.Split(string(jsonBytes), "\n") {
+		jsonBuilder.WriteString(line)
+		if strings.HasPrefix(line, "}") {
+			var mod listedModule
+			if err := json.Unmarshal([]byte(jsonBuilder.String()), &mod); err != nil {
+				return nil, err
+			}
+			jsonBuilder.Reset()
+			if mod.Path == "github.com/pingcap/tidb" {
+				continue
+			}
+			ret[mod.Path] = mod
+		}
+	}
+	return ret, nil
+}
+
+func getExistingMirrors() (map[string]DownloadableArtifact, error) {
+	depsbzl, err := bazel.Runfile("DEPS.bzl")
+	if err != nil {
+		return nil, err
+	}
+	return ListArtifactsInDepsBzl(depsbzl)
+}
+
+func mungeBazelRepoNameComponent(component string) string {
+	component = strings.ReplaceAll(component, "-", "_")
+	component = strings.ReplaceAll(component, ".", "_")
+	return strings.ToLower(component)
+}
+
+func modulePathToBazelRepoName(mod string) string {
+	components := strings.Split(mod, "/")
+	head := strings.Split(components[0], ".")
+	for i, j := 0, len(head)-1; i < j; i, j = i+1, j-1 {
+		head[i], head[j] = mungeBazelRepoNameComponent(head[j]), mungeBazelRepoNameComponent(head[i])
+	}
+	for index, component := range components {
+		if index == 0 {
+			continue
+		}
+		components[index] = mungeBazelRepoNameComponent(component)
+	}
+	return strings.Join(append(head, components[1:]...), "_")
+}
+
+func dumpPatchArgsForRepo(repoName string) error {
+	runfiles, err := bazel.RunfilesPath()
+	if err != nil {
+		return err
+	}
+	candidate := filepath.Join(runfiles, "build", "patches", repoName+".patch")
+	if _, err := os.Stat(candidate); err == nil {
+		if repoName == "io_etcd_go_etcd_api_v3" {
+			fmt.Printf(`        patch_args = ["-p2"],
+        patches = [
+            "//build/patches:%s.patch",
+        ],
+`, repoName)
+		} else {
+			fmt.Printf(`        patch_args = ["-p1"],
+        patches = [
+            "//build/patches:%s.patch",
+        ],
+`, repoName)
+		}
+	} else if !os.IsNotExist(err) {
+		return err
+	}
+	return nil
+}
+
+func buildFileProtoModeForRepo(repoName string) string {
+	if repoName == "io_etcd_go_etcd_api_v3" {
+		return "disable"
+	}
+	return "disable_global"
+}
+
+func dumpBuildNamingConventionArgsForRepo(repoName string) {
+	if repoName == "com_github_grpc_ecosystem_grpc_gateway" {
+		fmt.Printf("        build_naming_convention = \"go_default_library\",\n")
+	}
+}
+
+func dumpNewDepsBzl(
+	listed map[string]listedModule,
+	downloaded map[string]downloadedModule,
+	existingMirrors map[string]DownloadableArtifact,
+) error {
+	var sorted []string
+	repoNameToModPath := make(map[string]string)
+	for _, mod := range listed {
+		repoName := modulePathToBazelRepoName(mod.Path)
+		sorted = append(sorted, repoName)
+		repoNameToModPath[repoName] = mod.Path
+	}
+	sort.Strings(sorted)
+
+	ctx := context.Background()
+	var client *storage.Client
+	if canMirror() && isUpload {
```

Comment:
Why not just use `isMirror` just like `isUpload`?

Reply:
replace CanMirror with isMirror.